[Quirk] Allow users to upload HEIC images on zillow.com

karlcow opened the pull request:
#### ac5ea97ba27365aafaa79c15b47f935e076d2e51
<pre>
[Quirk] Allow users to upload HEIC images on zillow.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=292041">https://bugs.webkit.org/show_bug.cgi?id=292041</a>
<a href="https://rdar.apple.com/79872092">rdar://79872092</a>

Reviewed by Brent Fulgham.

When users of the zillow websites want to upload an HEIC images,
they get a message saying:

    HEIC format is not supported on web. Please download
    the Zillow Rental Manager App to upload, or convert
    the files to .png, .jpg, .jpeg, or .gif to upload.

WebKit added the possibility to automatically convert the HEIC image
format to one that the site accepts.
<a href="https://github.com/WebKit/WebKit/commit/1fd3fe3f407ad3cbc9fde83b03a9a0641e6d3e58">https://github.com/WebKit/WebKit/commit/1fd3fe3f407ad3cbc9fde83b03a9a0641e6d3e58</a>

The current input on the zillow website is

    &lt;input accept=&quot;image/*&quot; multiple=&quot;&quot; type=&quot;file&quot;&gt;

To benefit from WebKit capability to auto-convert the file
to the format of website choice, it is necessary to specify
in the accept attribute, the accepted format, for example:

    accept=&quot;image/jpeg, image/png, image/gif&quot;

This quirk is forcing for zillow image/jpeg as a conversion target.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldTranscodeHeicImagesForURL):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChooseFilesForOpenPanelWithImageTranscoding):

Canonical link: <a href="https://commits.webkit.org/294726@main">https://commits.webkit.org/294726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/744288cd20e58f1e7758a28a8b5f9b16abc571bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107748 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53224 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104626 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22564 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30762 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78036 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35012 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105593 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17481 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92809 "Found 1 new API test failure: TestWebKitAPI.SiteIsolation.RunOpenPanel (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58371 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17320 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10619 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52581 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87129 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10688 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110123 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29719 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87017 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30083 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88785 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86620 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31442 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9170 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23969 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16688 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29647 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34959 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29458 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32781 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31017 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10337,8 +10337,16 @@ bool WebPageProxy::didChooseFilesForOpenPanelWithImageTranscoding(const Vector<S
+        // For designated sites which are sending "image/*", we need to force the Mimetype to be able to transcode HEIC.
+        // FIXME: How to pass document->quirks() here?
+        if (document->quirks().needsForcingImageTranscoding())
+            transcodingMIMEType = "image/jpeg"_s;
```
This is not yet working. We need to have the context of the document or the URL to pass to the function… 
hmmm writing that I could probably just do that.

karlcow replied:
We passed currentURL()

karlcow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10337,8 +10337,16 @@ bool WebPageProxy::didChooseFilesForOpenPanelWithImageTranscoding(const Vector<S
     auto transcodingMIMEType = WebCore::MIMETypeRegistry::preferredImageMIMETypeForEncoding(allowedMIMETypes, { });
-    if (transcodingMIMEType.isNull())
-        return false;
+ß
```
hmm typo.

karlcow replied:
fixed

brentfulgham commented on Source/WebCore/page/Quirks.h:
```diff
@@ -136,6 +136,9 @@ class Quirks {
     static bool shouldMakeEventListenerPassive(const EventTarget&, const EventTypeInfo&);
 
+    WEBCORE_EXPORT static bool shouldTranscodeHeicImagesForURL(const URL&);
+
```
Nit: Only one newline please.

karlcow replied:
fixed locally.

brentfulgham commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -10335,8 +10335,15 @@ bool WebPageProxy::didChooseFilesForOpenPanelWithImageTranscoding(const Vector<S
+    if (transcodingMIMEType.isNull()) {
+        // For designated sites which are sending "image/*", we need to force the mimetype
+        // to be able to transcode from HEIC to JPEG.
+        if (Quirks::shouldTranscodeHeicImagesForURL(currentURL()))
```
The UIProcess side of things doesn't have a document object, so you have to check for the need to use quirks at the Page Settings level, like we do elsewhere in `WebPageProxy.cpp`.

Something like this:

```
        if (transcodingMIMEType.isNull() && protectedPreferences()->needsSiteSpecificQuirks()) {
            if (Quirks::shouldTranscodeHeicImagesForURL(currentURL()))
...
```

karlcow replied:
I think the code needs to separate the two cases of isNull and needsSiteSpecificQuirks.

karlcow replied:
New PR.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -703,6 +703,21 @@ bool Quirks::shouldAvoidScrollingWhenFocusedContentIsVisible() const
+// which do not support HEIC.
+bool Quirks::shouldTranscodeHeicImagesForURL(const URL& url)
+{
+    if (!needsQuirks())
```
This is a `static` method on the class. It doesn't have any 'state', and doesn't have a Document to ask.

In these cases, we have to check (in the UIProcess side) the page settings `protectedPreferences()->needsSiteSpecificQuirks()` as I showed earlier in this review.

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -703,6 +703,21 @@ bool Quirks::shouldAvoidScrollingWhenFocusedContentIsVisible() const
+    auto host = url.host();
+    // zillow.com rdar://79872092
+    if (host == "zillow.com"_s)
+        return true;
```
This is probably fine for `Zillow.com`, which is a simple URL. If we wanted to check more complicated things we would want to use:
` auto quirksDomainString = RegistrableDomain(quirksURL).string();`

karlcow replied:
ok modified.

brentfulgham reviewed (changes requested):
This looks really close! But you need to member that your check in this Quirk is happening on the UIProcess side (hence WebPageProxy instead of WebPage) so it doesn't have a Document to check. Instead, we have to look at the page settings to decide if Quirks are enabled.

karlcow commented on Source/WebCore/page/QuirksData.h:
```diff
@@ -52,6 +52,7 @@ struct WEBCORE_EXPORT QuirksData {
     bool needsBodyScrollbarWidthNoneDisabledQuirk : 1 { false };
     bool needsCanPlayAfterSeekedQuirk : 1 { false };
     bool needsChromeMediaControlsPseudoElementQuirk : 1 { false };
+    bool needsForcingImageTranscodingQuirk : 1 { false };
```
I need to remove this.

brentfulgham reviewed (approved):
I think this looks good. Let's see if rerunning the API tests clears up the issue.

karlcow committed:
[Quirk] Allow users to upload HEIC images on zillow.com
https://bugs.webkit.org/show_bug.cgi?id=292041
rdar://79872092

Reviewed by Brent Fulgham.

When users of the zillow websites want to upload an HEIC images,
they get a message saying:

    HEIC format is not supported on web. Please download
    the Zillow Rental Manager App to upload, or convert
    the files to .png, .jpg, .jpeg, or .gif to upload.

WebKit added the possibility to automatically convert the HEIC image
format to one that the site accepts.
https://github.com/WebKit/WebKit/commit/1fd3fe3f407ad3cbc9fde83b03a9a0641e6d3e58

The current input on the zillow website is

    <input accept="image/*" multiple="" type="file">

To benefit from WebKit capability to auto-convert the file
to the format of website choice, it is necessary to specify
in the accept attribute, the accepted format, for example:

    accept="image/jpeg, image/png, image/gif"

This quirk is forcing for zillow image/jpeg as a conversion target.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldTranscodeHeicImagesForURL):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didChooseFilesForOpenPanelWithImageTranscoding):

Canonical link: https://commits.webkit.org/294726@main
